libtsuba: Random fault injection tests

witchel opened the pull request:
This PR still has issues, but I want to put something on the board and this code does crash updates and verify the resulting graph in a useful way.  I'll open issues that are limitations, the big one being the files it writes are smaller than 8KB, leaving it out of most interesting S3 write paths.

I'm not sure why generate_random_alphanumeric_string is snake case, but I followed suit.

ddn0 commented on libsupport/include/galois/Random.h:
```diff
@@ -6,6 +6,12 @@
 
 std::string generate_random_alphanumeric_string(uint64_t len);
+// Range 0..len-1, inclusive
+int64_t generate_uniform_int(int64_t len);
```
Ugh. Not your issue since you are just following the style in this file, but since libsupport is new code, it should follow `PascalCase` for function names.

If you don't want to fix this in passing, can you file an issue so we remember to fix this?

ddn0 commented on libtsuba/tests/tsuba_fault.cpp:
```diff
@@ -0,0 +1,332 @@
+// During a graph mutation, crash horribly
+#include <time.h>
```
`#include <ctime>`

witchel commented:
> Seems reasonable to me.
> 
> My natural inclination would be to push the pull the plug calls at least one level lower, e.g., all Persist(), Barrier(), etc calls can fail at their beginning and end. We would lose the ability to be more fine-grained about fault sensitivity and reflecting fault boundaries explicitly for callers, but the application of fault injection would be more consistent.
> 
> I suppose a more generic syscall interposition is possible but at the expense of a much larger search space?

Well, Persist() calls end up calling into s3.cpp where I put PtP calls in particularly vulnerable places.  I assume you are not requesting that I take those PtP calls out?  Because those seem like the most valuable.

I want to limit the number of files I pollute, and RDG is attractive because it has high-level (from the storage perspective) perspective on what it is writing and why.   It calls a variety of storage interfaces, so the benefit of placing the PtP calls here is their focus and the "documentation" that the context of RDG provides.  But if you want to point me somewhere else, I'll listen.

One next step is a signal-based killing mechanism to do more random testing.  I suppose we could also look at per-syscall crashes, especially if you have an easy way of getting access to syscalls via c++.  If you are thinking something like ptrace to interpose on syscalls, I could look into it.

ddn0 commented:
> Well, Persist() calls end up calling into s3.cpp where I put PtP calls in particularly vulnerable places. I assume you are not requesting that I take those PtP calls out? Because those seem like the most valuable.

I was suggesting that the PtP calls that are currently before and after (every?) `ff->Persist` call might just be put inside `Persist` itself.

>  If you are thinking something like ptrace to interpose on syscalls, I could look into it.

I don't think syscall/ptrace is necessary now unless there was some just-works solution you knew off.

> I want to limit the number of files I pollute, and RDG is attractive because it has high-level (from the storage perspective) perspective on what it is writing and why. It calls a variety of storage interfaces, so the benefit of placing the PtP calls here is their focus and the "documentation" that the context of RDG provides.

My thinking is that the minimal number of PtP calls would be at each interesting point at the lowest level that we choose to instrument.

What I think I hear is that there is also value in putting PtP calls at the highest level of abstraction (RDG) for documentation and to test failures in contextually interesting areas, and putting PtP calls below RDG kind of defeats the point of these additional PtP calls. And this is why putting PtP in `Persist` (which is below the highest level and above the lowest) doesn't make sense.

This makes sense to me.

witchel commented:
> What I think I hear is that there is also value in putting PtP calls at the highest level of abstraction (RDG) for documentation and to test failures in contextually interesting areas, and putting PtP calls below RDG kind of defeats the point of these additional PtP calls. And this is why putting PtP in `Persist` (which is below the highest level and above the lowest) doesn't make sense.
> 
> This makes sense to me.

This is what I wish I had said.

witchel commented:
I want to put this to bed, but two notes.
1. There was a "disagreement in semantics" between my input graphml file and @patrickkenney9801 's converter script.  I modified my input and the crash problem I saw went away.
2. I have introduced a call to __builtin_return_address, which is a gcc dependence.  @ddn0 , is there some way I should guard such code?  The alternative would be to pass __FILE__ and __LINE__ into PtP, which seems like it would make PtP even uglier.

ddn0 commented:
> I have introduced a call to __builtin_return_address, which is a gcc dependence. @ddn0 , is there some way I should guard such code? The alternative would be to pass FILE and LINE into PtP, which seems like it would make PtP even uglier.

Well there is the proper way (e.g., add a macro in libsupport `Platform.h` with some configure time checks to see if the current compiler supports it), and there is the expedient way, `__builtin_return_address` seems to be supported by LLVM and GCC for the versions we care about so punt this to a time when someone reports an issue.

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,202 @@
+
+namespace {
+
+std::vector<std::string>
```
We have `katana::SplitView` in `Strings.h` and, in general, `Strings.h` is the place for basic string utilities when you don't really care about large inputs, etc but just want convenience, which I think applies here.

arthurp replied:
Oops. 👍🏻

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  std::vector<katana::Plugin> plugins() {
```
`const`?

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,202 @@
+};
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+PluginLoader plugin_loader;
```
Globals are named like constants:

https://google.github.io/styleguide/cppguide.html#Constant_Names

arthurp replied:
Having the `k` on it seems like a bad idea since this isn't a const (The `k` is for konst IIUC). So I'm renaming to `PluginLoader`. Is that OK?

ddn0 replied:
I take this back, I think `plugin_loader` is an ok name according to the style guide.

ddn0 commented on libsupport/include/katana/Plugin.h:
```diff
@@ -0,0 +1,99 @@
+#include <boost/filesystem.hpp>
+#include <katana/config.h>
+
+/**
```
nit: `///` since the rest of the file uses that style for doxygen comments.

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,187 @@
+
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
```
maybe follow the initial case of other log messages (lower case)

arthurp replied:
OK. I thought error messages where initial lowercase and log messages where initial capitol because they were sentences. I'll change this and in other places.

ddn0 replied:
I actually don't know if we have a convention for log messages. We should probably pick one. I was more reacting to the varying capitialization in the log, warn and warn_once messages in this file.

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,187 @@
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
+      for (const boost::filesystem::directory_entry& entry :
```
- does this iterator guarantee an order in entries?
- do we care if plugins are loaded in arbitrary order?

arthurp replied:
The order is arbitrary (probably usually FS level dir entry order). I don't want to guarantee any plugin loading order yet. If we want to add that later we can. Doing it right would probably require dependency information or a way to directly control loading. Right now I'm intentionally being dumb and just loading everything that can be loaded. This defers decisions about what plugins are allowed to do.

Do you think this is a mistake?

ddn0 replied:
Doing dependencies seems like a lot of work. I was more wondering if we want the order to be deterministic or if we want to intentionally randomize the order to prevent people from depending on "natural" filesystem order.

arthurp replied:
Yea. I would be temped to randomize to prevent people from making assumptions or naming plugins to cause a specific order (if we sorted by file name). I'll add randomization. Preventing assumptions now will be useful so we can add features later without needing to be backwards compatible with some early assumptions.

ddn0 commented on libsupport/src/Plugin.cpp:
```diff
@@ -0,0 +1,187 @@
+      if (!err) {
+        // 2a) installed shared objects
+        auto path = canonical(
+            this_so_filename.parent_path() / "katana" / "plugins", err);
```
can we dump plugins to CMAKE_BINARY_PATH or something so we don't need to hardcode the relation of different build directories?

arthurp replied:
Yea. This is a hack. Putting them in `CMAKE_BINARY_PATH` wouldn't help since this loading doesn't use `LD_LIBRARY_PATH` or rpaths. I did this because it means I don't need to change all the launchers, but I could remove this and instead set `KATANA_PLUGIN_PATH` in the ctest config and is `python_env.sh`. What's your preference?

ddn0 replied:
I don't think `python_env.sh` would work that well because we'd have to train people to use that always.

Nothing obvious comes up to me for cleaning this up.

Though, perhaps we should be more strict about the paths we check: (1) if KATANA_PLUGIN_PATH is set use it and no other; otherwise (2) if installed plugin directory exists, use it and no other; finally (3) try only build-tree based paths.

arthurp replied:
I actually already do 1 and 2. I will add 3 as conditional step. I will also add a WARNING that is printed when build tree plugins are used, so people don't get surprised by it.

And yes, I just realized the issue. People obviously run native programs directly. Don't know how I forgot that.

ddn0 commented on libtsuba/src/RDGMeta.h:
```diff
@@ -5,6 +5,8 @@
 #include <regex>
 #include <set>
 
+#include <tsuba/NameServerClient.h>
```
`#include "tsuba/NameServerClient.h"` probably

tylershunt reviewed (approved):
LGTM! thanks!
